### PR TITLE
fix(gpt2): Resolve NaN/Inf issue in lm_head on Python 3.13 with tied weights

### PR DESCRIPTION
**Pull request by JokeYoonic**
Problem:
- On macOS ARM64 + Python 3.13 + transformers 5.x, GPT-2 model's lm_head forward pass produces NaN/Inf values during inference
- Root cause: lm_head.weight is tied to transformer.wte.weight, and the shared memory reference causes numerical instability in Python 3.13

Solution:
- Clone the lm_head weight before passing to F.linear in GPT2LMHeadModel and GPT2DoubleHeadsModel forward methods
- This breaks the memory sharing and resolves the NaN issue

Changes:
- src/transformers/models/gpt2/modeling_gpt2.py: Modified GPT2LMHeadModel.forward() and GPT2DoubleHeadsModel.forward() to use self.lm_head.weight.clone()

Testing:
- Verified fix with gpt2-medium model on Python 3.13.5 + PyTorch 2.6.0
- All existing GPT-2 model tests pass
## Reproduction Code

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model = AutoModelForCausalLM.from_pretrained("gpt2-medium")
model.eval()
tokenizer = AutoTokenizer.from_pretrained("gpt2-medium")
inputs = tokenizer("Hello world", return_tensors="pt")

with torch.no_grad():
    outputs = model(**inputs)
    logits = outputs.logits

print(f"Has NaN: {torch.isnan(logits).any().item()}")  # Should be False after fix
print(f"Has Inf: {torch.isinf(logits).any().item()}")  # Should be False after fix
```

# What does this PR do?

Fixes # (issue)

## Before submitting
- [x ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by gambletan — COMMENTED** [review]
## Review: GPT-2 NaN/Inf fix via weight cloning

**Overall concern: This fix addresses a symptom rather than the root cause, and introduces a performance regression on every forward pass.**

### Issues:

1. **Performance impact — `.clone()` on every forward pass:** `self.lm_head.weight.clone()` creates a full copy of the weight matrix on every single forward call. For `gpt2-medium`, `lm_head.weight` is shape `(50257, 1024)` — that's ~200MB of extra memory allocation and copy per forward pass (in float32). For `gpt2-xl` it would be even larger. This is a significant overhead in both memory and compute, especially for inference serving where you want minimal allocation. If this fix is truly needed, it should at minimum be guarded by a condition (e.g., only clone when weights are actually tied, or only on affected Python versions).

2. **Root cause is unclear:** The PR claims "shared memory reference causes numerical instability in Python 3.13" but this is a very unusual claim. Python version changes should not affect PyTorch tensor numerical behavior — the computation happens in C++/CUDA kernels, not in the Python interpreter. This suggests the real bug is elsewhere. Possible alternative explanations:
   - A PyTorch version difference that happened to coincide with the Python 3.13 upgrade
   - A change in how `nn.Linear` handles tied weights in a specific PyTorch build
   - An issue with the specific build of NumPy/PyTorch compiled for Python 3.13
   
   Without a linked issue (`https://github.com/huggingface/transformers/issues/XXXXX` is still a placeholder), it's hard to assess.

3. **The placeholder issue link (`issues/XXXXX`):** The comment references a non-existent issue number. This should be updated to point to an actual issue, or removed.

4. **This fix is GPT-2-specific but the problem (if real) would affect all models with tied weights:** Every model that uses `tie_word_embeddings=True` shares the same weight tensor between the embedding and the LM head. If this were truly a Python 3.13 numerical stability bug with tied weights, it would affect LLaMA, GPT-J, Falcon, and dozens of other models. A fix scoped only to GPT-2 suggests either (a) the root cause is GPT-2-specific (unlikely for a weight-tying issue) or (b) the diagnosis is incomplete.

5. **No test added:** There's no regression test to verify the NaN/Inf behavior is fixed. Even a simple test that runs a forward pass and asserts `not torch.isnan(logits).any()` would help ensure this doesn't regress.

### Recommendation:
Before merging, I'd suggest:
- Confirming this reproduces with the *same* PyTorch version on both Python 3.12 and 3.13 to rule out PyTorch build differences
- If the `.clone()` fix is truly necessary, guard it with `if self.lm_head.weight.data_ptr() == self.transformer.wte.weight.data_ptr()` to only clone when weights are actually tied
- File an actual issue and link it

**Comment by Rocketknight1** [comment]
Hi, "shared memory reference causes numerical instability" doesn't make any sense to me, because "numerical instability" usually refers to errors in the least significant bits caused by floating point precision. Is the tied weight  just not being handled correctly and we're getting randomly-initialized values instead?
